p-ata/src/processor.rs

```diff
@@ -0,0 +1,595 @@
+        len if len >= 7 => Some(unsafe { accounts.get_unchecked(6) }),
+        6 => None,
+        _ => return Err(ProgramError::NotEnoughAccountKeys),
+    };
+
+    // SAFETY: account len already checked
+    unsafe {
+        Ok(CreateAccounts {
+            payer: accounts.get_unchecked(0),
+            associated_token_account_to_create: accounts.get_unchecked(1),
+            wallet: accounts.get_unchecked(2),
+            mint: accounts.get_unchecked(3),
+            system_program: accounts.get_unchecked(4),
+            token_program: accounts.get_unchecked(5),
+            rent_sysvar: rent_info,
+        })
+    }
+}
+
+/// Check if account already exists and is properly configured (idempotent check).
+///
+/// This function validates that an existing ATA account:
+/// 1. Is owned by the correct token program
+/// 2. Has the correct owner (wallet)
+/// 3. Has the correct mint
+/// 4. Uses the canonical PDA address, even when `expected_bump` is provided
+///
+/// ## Returns
+///
+/// * `Ok(true)` - Account exists and is properly configured (safe to skip creation)
+/// * `Ok(false)` - Account doesn't exist or is not token program owned (needs creation)  
+/// * `Err(_)` - Account exists but has invalid configuration (error condition)
+#[inline(always)]
+pub(crate) fn check_idempotent_account(
+    associated_token_account: &AccountInfo,
+    wallet: &AccountInfo,
+    mint_account: &AccountInfo,
+    token_program: &AccountInfo,
+    idempotent: bool,
+    program_id: &Pubkey,
+    expected_bump: Option<u8>,
+) -> Result<bool, ProgramError> {
+    if idempotent && associated_token_account.is_owned_by(token_program.key()) {
+        let ata_state = get_token_account(associated_token_account)?;
+
+        validate_token_account_owner(ata_state, wallet.key())?;
+        validate_token_account_mint(ata_state, mint_account.key())?;
+
+        match expected_bump {
+            Some(bump) => {
+                let seeds: &[&[u8]; 3] = &[
+                    wallet.key().as_ref(),
+                    token_program.key().as_ref(),
+                    mint_account.key().as_ref(),
+                ];
+                let (verified_address, verified_bump) =
+                    ensure_no_better_canonical_address_and_bump(seeds, program_id, bump);
+
+                let maybe_canonical_address = verified_address
+                    .unwrap_or_else(|| derive_address::<3>(seeds, Some(verified_bump), program_id));
+
+                // We must check that the actual derived address is off-curve,
+                // since it will not fail downstream as in Create paths.
+                // Potential problem if skipping this is demonstrated in
+                // tests/test_mollusk_oncurve_attack.rs
+                if !is_off_curve(&maybe_canonical_address)
+                    || maybe_canonical_address != *associated_token_account.key()
+                {
+                    msg!("PROGRAM: addresses don't match");
```

Comment:
nit: any reason for the `PROGRAM:` prefix in the log message? If not, best to remove, or even better, give more info with those bytes! ie `Provided bump {bump} does not result in canonical or provided address`